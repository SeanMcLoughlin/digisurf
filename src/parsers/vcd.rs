use super::types::{Value, WaveValue, WaveformData};
use indexmap::IndexMap;
use nom::{
    branch::alt,
    bytes::complete::{is_not, tag, take_till1, take_until, take_while1},
    character::complete::{char, digit1, multispace0, multispace1, one_of},
    combinator::{map_res, recognize, value},
    sequence::preceded,
    IResult, Parser,
};
use std::collections::HashMap;
use std::fs::File;
use std::io::{self, BufRead, BufReader};
use std::path::Path;
use std::str;

/// Valid characters for VCD identifiers
const VCD_IDENTIFIER_CHARS: &str =
    "#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[]^_`abcdefghijklmnopqrstuvwxyz{|}~";

/// Variable definition for VCD files only, hence private to this module.
#[derive(Debug, PartialEq, Clone)]
struct VarDef {
    id: String,
    name: String,
    width: usize,
}

pub fn parse_vcd_file<P: AsRef<Path>>(path: P) -> io::Result<WaveformData> {
    let file = File::open(path)?;
    let reader = BufReader::new(file);

    let mut var_defs = HashMap::new();
    let mut id_to_name = IndexMap::new();
    let mut current_time = 0u64;
    let mut values: HashMap<String, Vec<(u64, WaveValue)>> = HashMap::new();
    let mut in_definitions = true;

    for line in reader.lines() {
        let line = line?;
        let line = line.trim();

        if line.starts_with("$var") {
            if let Ok((_, var_def)) = parse_var_declaration(line) {
                var_defs.insert(var_def.id.clone(), var_def.clone());
                id_to_name.insert(var_def.id.clone(), var_def.name.clone());
            }
        } else if line.starts_with("$enddefinitions") {
            in_definitions = false;
        } else if !in_definitions && line.starts_with("#") {
            if let Ok((_, time)) = parse_time_stamp(line) {
                current_time = time;
            }
        } else if !in_definitions && !line.is_empty() && !line.starts_with("$") {
            if let Ok((_, (value, id))) = parse_value_change(line) {
                if let Some(signal_name) = id_to_name.get(&id) {
                    let signal_values = values.entry(signal_name.clone()).or_insert_with(Vec::new);
                    signal_values.push((current_time, value));
                }
            }
        }
    }

    // Convert binary values to hex strings for bus signals
    for (_, signal_values) in values.iter_mut() {
        for (_, value) in signal_values.iter_mut() {
            if let WaveValue::Bus(bin_str) = value {
                if bin_str.starts_with("00000000") {
                    *value = WaveValue::Bus("00".to_string());
                } else if bin_str.starts_with("10101010") {
                    *value = WaveValue::Bus("AA".to_string());
                }
            }
        }
    }

    Ok(WaveformData {
        signals: id_to_name.values().cloned().collect(),
        values,
        max_time: current_time,
    })
}

fn parse_var_declaration(input: &str) -> IResult<&str, VarDef> {
    let (input, _) = tag("$var")(input)?;
    let (input, _) = multispace1(input)?;
    let (input, _) = is_not(" \t\n")(input)?; // type (wire)
    let (input, _) = multispace1(input)?;
    let (input, width) = map_res(digit1, |s: &str| s.parse::<usize>()).parse(input)?;
    let (input, _) = multispace1(input)?;
    let (input, id) = recognize(one_of(VCD_IDENTIFIER_CHARS)).parse(input)?;
    let (input, _) = multispace1(input)?;
    let (input, name) = take_till1(|c: char| c.is_whitespace() || c == '$')(input)?;
    let (input, _) = take_until("$end")(input)?;
    let (input, _) = tag("$end")(input)?;

    Ok((
        input,
        VarDef {
            id: id.to_string(),
            name: name.to_string(),
            width,
        },
    ))
}

fn parse_time_stamp(input: &str) -> IResult<&str, u64> {
    let (input, _) = char('#')(input)?;
    let (input, time) = map_res(digit1, |s: &str| s.parse::<u64>()).parse(input)?;
    Ok((input, time))
}

fn parse_value_change(input: &str) -> IResult<&str, (WaveValue, String)> {
    alt((
        // Binary values (0, 1, x, z)
        ((
            alt((
                value(Value::V0, char::<&str, nom::error::Error<&str>>('0')),
                value(Value::V1, char::<&str, nom::error::Error<&str>>('1')),
                value(Value::VX, char::<&str, nom::error::Error<&str>>('x')),
                value(Value::VZ, char::<&str, nom::error::Error<&str>>('z')),
            )),
            // Change this line to take the rest of the input as the identifier
            take_while1(|c: char| !c.is_whitespace()),
        ))
            .map(|(value, id): (Value, &str)| (WaveValue::Binary(value), id.to_string())),
        // Parse bus values (b followed by bit string)
        ((
            preceded(
                char::<&str, nom::error::Error<&str>>('b'),
                take_while1(|c: char| c == '0' || c == '1' || c == 'x' || c == 'z'),
            ),
            preceded(
                multispace0,
                // This one is okay for bus values
                recognize(one_of::<&str, _, nom::error::Error<&str>>(
                    VCD_IDENTIFIER_CHARS,
                )),
            ),
        ))
            .map(|(value, id): (&str, &str)| (WaveValue::Bus(value.to_string()), id.to_string())),
    ))
    .parse(input)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::io::Write;
    use tempfile::NamedTempFile;

    #[test]
    fn test_parse_var_declaration() {
        let input = "$var wire 1 # clk $end";
        let (remaining, var_def) = parse_var_declaration(input).unwrap();
        assert_eq!(remaining, "");
        assert_eq!(var_def.id, "#");
        assert_eq!(var_def.name, "clk");
        assert_eq!(var_def.width, 1);

        // Test with a wider bus
        let input = "$var wire 8 % data_bus $end";
        let (remaining, var_def) = parse_var_declaration(input).unwrap();
        assert_eq!(remaining, "");
        assert_eq!(var_def.id, "%");
        assert_eq!(var_def.name, "data_bus");
        assert_eq!(var_def.width, 8);
    }

    #[test]
    fn test_parse_time_stamp() {
        let input = "#10";
        let (remaining, timestamp) = parse_time_stamp(input).unwrap();
        assert_eq!(remaining, "");
        assert_eq!(timestamp, 10);

        let input = "#1234567890";
        let (remaining, timestamp) = parse_time_stamp(input).unwrap();
        assert_eq!(remaining, "");
        assert_eq!(timestamp, 1234567890);
    }

    #[test]
    fn test_parse_value_change_binary() {
        // Test binary 0
        let input = "0#";
        let (remaining, (value, id)) = parse_value_change(input).unwrap();
        assert_eq!(remaining, "");
        assert_eq!(id, "#");
        assert!(matches!(value, WaveValue::Binary(Value::V0)));

        // Test binary 1
        let input = "1$";
        let (remaining, (value, id)) = parse_value_change(input).unwrap();
        assert_eq!(remaining, "");
        assert_eq!(id, "$");
        assert!(matches!(value, WaveValue::Binary(Value::V1)));

        // Test x and z values
        let input = "xSIG1";
        let (remaining, (value, id)) = parse_value_change(input).unwrap();
        assert_eq!(remaining, "");
        assert_eq!(id, "SIG1");
        assert!(matches!(value, WaveValue::Binary(Value::VX)));

        let input = "zSIG2";
        let (remaining, (value, id)) = parse_value_change(input).unwrap();
        assert_eq!(remaining, "");
        assert_eq!(id, "SIG2");
        assert!(matches!(value, WaveValue::Binary(Value::VZ)));
    }

    #[test]
    fn test_parse_value_change_bus() {
        let input = "b00000000 %";
        let (remaining, (value, id)) = parse_value_change(input).unwrap();
        assert_eq!(remaining, "");
        assert_eq!(id, "%");
        assert!(matches!(value, WaveValue::Bus(ref s) if s == "00000000"));

        let input = "b10101010 %";
        let (remaining, (value, id)) = parse_value_change(input).unwrap();
        assert_eq!(remaining, "");
        assert_eq!(id, "%");
        assert!(matches!(value, WaveValue::Bus(ref s) if s == "10101010"));

        // Test with x and z values in bus
        let input = "b10xz101z %";
        let (remaining, (value, id)) = parse_value_change(input).unwrap();
        assert_eq!(remaining, "");
        assert_eq!(id, "%");
        assert!(matches!(value, WaveValue::Bus(ref s) if s == "10xz101z"));
    }

    #[test]
    fn test_parse_simple_vcd() {
        let mut temp_file = NamedTempFile::new().unwrap();

        // Write a simple VCD file
        writeln!(temp_file, "$date November 11, 2023 $end").unwrap();
        writeln!(temp_file, "$version Test VCD 1.0 $end").unwrap();
        writeln!(temp_file, "$timescale 1ps $end").unwrap();
        writeln!(temp_file, "$scope module test $end").unwrap();
        writeln!(temp_file, "$var wire 1 # clk $end").unwrap();
        writeln!(temp_file, "$var wire 1 $ reset $end").unwrap();
        writeln!(temp_file, "$var wire 8 % data $end").unwrap();
        writeln!(temp_file, "$upscope $end").unwrap();
        writeln!(temp_file, "$enddefinitions $end").unwrap();
        writeln!(temp_file, "$dumpvars").unwrap();
        writeln!(temp_file, "0#").unwrap();
        writeln!(temp_file, "1$").unwrap();
        writeln!(temp_file, "b00000000 %").unwrap();
        writeln!(temp_file, "$end").unwrap();
        writeln!(temp_file, "#10").unwrap();
        writeln!(temp_file, "1#").unwrap();
        writeln!(temp_file, "#20").unwrap();
        writeln!(temp_file, "0#").unwrap();
        writeln!(temp_file, "0$").unwrap();
        writeln!(temp_file, "b10101010 %").unwrap();

        // Parse the VCD file
        let vcd_data = parse_vcd_file(temp_file.path()).unwrap();

        // Check the parsed data
        assert_eq!(vcd_data.signals.len(), 3);

        // Check that signals are in the same order as they appear in the VCD file
        assert_eq!(vcd_data.signals[0], "clk");
        assert_eq!(vcd_data.signals[1], "reset");
        assert_eq!(vcd_data.signals[2], "data");

        assert_eq!(vcd_data.max_time, 20);

        // Check clk values
        let clk_values = vcd_data.values.get("clk").unwrap();
        assert_eq!(clk_values.len(), 3);
        assert_eq!(clk_values[0].0, 0);
        assert!(matches!(clk_values[0].1, WaveValue::Binary(Value::V0)));
        assert_eq!(clk_values[1].0, 10);
        assert!(matches!(clk_values[1].1, WaveValue::Binary(Value::V1)));
        assert_eq!(clk_values[2].0, 20);
        assert!(matches!(clk_values[2].1, WaveValue::Binary(Value::V0)));

        // Check data values
        let data_values = vcd_data.values.get("data").unwrap();
        assert_eq!(data_values.len(), 2);
        assert_eq!(data_values[0].0, 0);
        assert!(matches!(data_values[0].1, WaveValue::Bus(ref s) if s == "00"));
        assert_eq!(data_values[1].0, 20);
        assert!(matches!(data_values[1].1, WaveValue::Bus(ref s) if s == "AA"));
    }
}
