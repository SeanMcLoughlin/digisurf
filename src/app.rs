use crate::{
    command_mode::CommandModeWidget,
    commands, constants,
    input::{COMMAND_KEYBINDINGS, NORMAL_KEYBINDINGS},
    state::AppState,
    types::{AppMode, WaveValue},
    ui::{
        layout::{create_layout, AppLayout},
        widgets::{
            bottom_text_box::BottomTextBoxWidget, help_menu::HelpMenuWidget,
            signal_list::SignalListWidget, title_bar::TitleBarWidget, waveform::WaveformWidget,
        },
    },
};
use crossterm::event::{
    self, Event, KeyCode, KeyModifiers, MouseButton, MouseEvent, MouseEventKind,
};
use ratatui::{
    layout::Rect,
    prelude::*,
    widgets::{Paragraph, Widget},
    DefaultTerminal,
};
use std::{error::Error, time::Duration};
use vcd::Value;

pub struct App {
    pub state: AppState,
    pub layout: AppLayout,
    pub signal_list: SignalListWidget,
    pub help_menu: HelpMenuWidget,
    pub waveform: WaveformWidget,
    pub title_bar: TitleBarWidget,
    pub command_input: BottomTextBoxWidget,
    pub command_result: Option<(String, bool)>, // (message, is_error)
    pub command_result_time: Option<std::time::Instant>,
    pub command_mode: CommandModeWidget<AppState>,
}

impl Default for App {
    fn default() -> Self {
        let mut app = App {
            state: AppState::new(),
            layout: AppLayout::default(),
            signal_list: SignalListWidget::default(),
            help_menu: HelpMenuWidget::default(),
            waveform: WaveformWidget::default(),
            title_bar: TitleBarWidget::default(),
            command_input: BottomTextBoxWidget::default(),
            command_result: None,
            command_result_time: None,
            command_mode: CommandModeWidget::new(),
        };
        app.generate_test_data();
        app.register_commands();
        app
    }
}

impl App {
    pub fn run(mut self, mut terminal: DefaultTerminal) -> Result<(), Box<dyn Error>> {
        let tick_rate = Duration::from_millis(250);
        while !self.state.exit {
            terminal.draw(|frame| frame.render_widget(&mut self, frame.area()))?;

            if event::poll(tick_rate)? {
                match event::read()? {
                    Event::Key(key) => {
                        if self.state.show_help {
                            match key.code {
                                KeyCode::Esc => {
                                    self.state.show_help = false;
                                    self.state.help_menu_scroll = 0;
                                }
                                KeyCode::Up => {
                                    if self.state.help_menu_scroll > 0 {
                                        self.state.help_menu_scroll -= 1;
                                    }
                                }
                                KeyCode::Down => {
                                    self.state.help_menu_scroll += 1;
                                }
                                _ => {}
                            }
                        } else if self.state.mode == AppMode::Command {
                            self.handle_command_input(key.code);
                        } else {
                            self.handle_input(key.code);
                        }
                    }
                    Event::Mouse(mouse) => self.handle_mouse(mouse),
                    _ => {}
                }
            } else {
                // Check if command result should be hidden
                if let Some(time) = self.command_result_time {
                    if time.elapsed().as_secs() >= constants::COMMAND_RESULT_HIDE_THRESHOLD_SECONDS
                    {
                        self.command_result = None;
                        self.command_result_time = None;
                    }
                }
            }
        }
        Ok(())
    }

    fn register_commands(&mut self) {
        commands::register_all_commands(&mut self.command_mode);
    }

    pub fn handle_command_input(&mut self, key: KeyCode) {
        match key {
            k if k == COMMAND_KEYBINDINGS.enter_normal_mode => {
                self.state.mode = AppMode::Normal;
                self.state.currently_typed_text_in_bottom_text_box.clear();
            }
            k if k == COMMAND_KEYBINDINGS.execute_command => {
                let command = self.state.currently_typed_text_in_bottom_text_box.clone();
                let result = self
                    .command_mode
                    .parser()
                    .execute(&command, &mut self.state);

                match result {
                    Ok(msg) => {
                        self.command_result = Some((msg, false));
                    }
                    Err(err) => {
                        self.command_result = Some((err, true));
                    }
                }
                self.command_result_time = Some(std::time::Instant::now());
                self.state.mode = AppMode::Normal;
                self.state.currently_typed_text_in_bottom_text_box.clear();
            }
            KeyCode::Backspace => {
                self.state.currently_typed_text_in_bottom_text_box.pop();
            }
            KeyCode::Char(c) => {
                self.state.currently_typed_text_in_bottom_text_box.push(c);
            }
            _ => {}
        }
    }

    pub fn handle_input(&mut self, key: KeyCode) {
        if self.state.mode == AppMode::Command {
            self.handle_command_input(key);
        } else {
            if key == NORMAL_KEYBINDINGS.enter_command_mode {
                self.state.mode = AppMode::Command;
                self.state.currently_typed_text_in_bottom_text_box.clear();
            } else {
                self.handle_normal_mode(key);
            }
        }
    }

    pub fn handle_mouse(&mut self, mouse: MouseEvent) {
        // Check if click is within waveform area
        if mouse.column >= self.layout.waveform.x
            && mouse.column <= self.layout.waveform.right()
            && mouse.row >= self.layout.waveform.y
            && mouse.row <= self.layout.waveform.bottom()
        {
            // Convert column to coordinates inside waveform area
            let column_in_waveform = mouse.column - self.layout.waveform.x;

            match mouse.kind {
                MouseEventKind::Down(MouseButton::Left) => {
                    if mouse.modifiers.contains(KeyModifiers::SHIFT) {
                        // Shift is still used for secondary marker
                        self.state
                            .set_secondary_marker(column_in_waveform, self.layout.waveform.width);
                    } else {
                        // Start potential drag or click - we don't know which yet
                        let time = self
                            .state
                            .screen_pos_to_time(column_in_waveform, self.layout.waveform.width);
                        self.state.drag_start = Some((column_in_waveform, time));
                        self.state.drag_current = Some((column_in_waveform, time));
                        self.state.is_dragging = false; // Not dragging yet
                    }
                }
                MouseEventKind::Drag(MouseButton::Left) => {
                    if self.state.drag_start.is_some() {
                        let time = self
                            .state
                            .screen_pos_to_time(column_in_waveform, self.layout.waveform.width);

                        if let Some((start_x, _)) = self.state.drag_start {
                            // Detect if we've moved enough to consider this a drag
                            if !self.state.is_dragging
                                && (start_x as i32 - column_in_waveform as i32).abs()
                                    >= constants::DRAG_DETECTED_THRESHOLD_PIXELS
                            {
                                self.state.is_dragging = true;
                            }
                        }

                        // Update current position regardless
                        self.state.drag_current = Some((column_in_waveform, time));
                    }
                }
                MouseEventKind::Up(MouseButton::Left) => {
                    if let (Some((start_x, start_time)), Some((end_x, end_time))) =
                        (self.state.drag_start, self.state.drag_current)
                    {
                        if self.state.is_dragging {
                            // This was a drag operation - zoom to selection
                            // Only zoom if dragged a minimum distance
                            if (start_x as i32 - end_x as i32).abs()
                                > constants::DRAG_STARTED_THRESHOLD_PIXELS
                            {
                                // Order the times correctly
                                let (min_time, max_time) = if start_time < end_time {
                                    (start_time, end_time)
                                } else {
                                    (end_time, start_time)
                                };

                                // Set the new zoom area
                                self.state.time_start = min_time;
                                self.state.time_range = max_time.saturating_sub(min_time).max(1);
                            }
                        } else {
                            // This was a click (not a drag) - set marker
                            self.state
                                .set_primary_marker(start_x, self.layout.waveform.width);
                        }
                    }

                    // Reset drag state
                    self.state.drag_start = None;
                    self.state.drag_current = None;
                    self.state.is_dragging = false;
                }
                _ => {}
            }
        } else {
            // Clear drag state if clicked outside the waveform area
            self.state.drag_start = None;
            self.state.drag_current = None;
            self.state.is_dragging = false;
        }
    }

    fn handle_normal_mode(&mut self, key: KeyCode) {
        match key {
            k if k == NORMAL_KEYBINDINGS.down => {
                self.state.selected_signal =
                    (self.state.selected_signal + 1) % self.state.signals.len();
            }
            k if k == NORMAL_KEYBINDINGS.up => {
                if self.state.selected_signal > 0 {
                    self.state.selected_signal -= 1;
                } else {
                    self.state.selected_signal = self.state.signals.len() - 1;
                }
            }
            k if k == NORMAL_KEYBINDINGS.left => {
                if self.state.time_start > 0 {
                    self.state.time_start = self
                        .state
                        .time_start
                        .saturating_sub(self.state.time_range / 4);
                }
            }
            k if k == NORMAL_KEYBINDINGS.right => {
                if self.state.time_start < self.state.max_time {
                    // Ensure the waveform view doesn't go beyond max_time
                    let max_start = self.state.max_time.saturating_sub(self.state.time_range);
                    self.state.time_start =
                        (self.state.time_start + self.state.time_range / 4).min(max_start);
                }
            }
            k if k == NORMAL_KEYBINDINGS.zoom_out => {
                // Calculate the new time range, doubling but capped at max_time
                let new_time_range = (self.state.time_range * 2).min(self.state.max_time);

                // Calculate center point of current view
                let center = self.state.time_start + (self.state.time_range / 2);

                // Calculate new start time, keeping the center point if possible
                let half_new_range = new_time_range / 2;
                let new_start = if center > half_new_range {
                    center.saturating_sub(half_new_range)
                } else {
                    0
                };

                // Make sure the end time (start + range) doesn't exceed max_time
                let adjusted_start = if new_start + new_time_range > self.state.max_time {
                    self.state.max_time.saturating_sub(new_time_range)
                } else {
                    new_start
                };

                self.state.time_start = adjusted_start;
                self.state.time_range = new_time_range;
            }
            k if k == NORMAL_KEYBINDINGS.zoom_in => {
                // Calculate center point of current view
                let center = self.state.time_start + (self.state.time_range / 2);

                // Calculate new time range, halving but with a minimum
                let new_time_range = (self.state.time_range / 2).max(10);

                // Calculate new start time, trying to keep the center point
                let half_new_range = new_time_range / 2;
                let new_start = center.saturating_sub(half_new_range);

                self.state.time_start = new_start;
                self.state.time_range = new_time_range;
            }
            k if k == NORMAL_KEYBINDINGS.zoom_full => {
                self.state.time_start = 0;
                self.state.time_range = self.state.max_time;
            }

            k if k == NORMAL_KEYBINDINGS.delete_primary_marker => {
                self.state.primary_marker = None;
            }
            k if k == NORMAL_KEYBINDINGS.delete_secondary_marker => {
                self.state.secondary_marker = None;
            }

            _ => {}
        }
    }

    pub fn generate_test_data(&mut self) {
        let test_signals = vec![
            "clk".to_string(),
            "reset".to_string(),
            "data_valid".to_string(),
            "data".to_string(),
            "tristate".to_string(),  // For Z states
            "undefined".to_string(), // For X states
        ];

        for signal in test_signals {
            self.state.signals.push(signal.clone());
            let mut values = Vec::new();

            for t in 0..100 {
                let value = match signal.as_str() {
                    "clk" => {
                        if t % 2 == 0 {
                            WaveValue::Binary(Value::V1)
                        } else {
                            WaveValue::Binary(Value::V0)
                        }
                    }
                    "reset" => {
                        if t < 10 {
                            WaveValue::Binary(Value::V1)
                        } else {
                            WaveValue::Binary(Value::V0)
                        }
                    }
                    "data_valid" => {
                        if t % 10 == 0 {
                            WaveValue::Binary(Value::V1)
                        } else {
                            WaveValue::Binary(Value::V0)
                        }
                    }
                    "data" => WaveValue::Bus(format!("{:02X}", t % 256)),
                    "tristate" => {
                        // Demonstrate high-impedance (Z) state every 3 cycles
                        if t % 3 == 0 {
                            WaveValue::Binary(Value::Z)
                        } else {
                            WaveValue::Binary(Value::V1)
                        }
                    }
                    "undefined" => {
                        // Demonstrate undefined (X) state every 5 cycles
                        if t % 5 == 0 {
                            WaveValue::Binary(Value::X)
                        } else {
                            WaveValue::Binary(Value::V0)
                        }
                    }
                    _ => WaveValue::Binary(Value::V0),
                };
                values.push((t as u64, value));
            }

            self.state.values.insert(signal, values);
        }
        self.state.max_time = 100;
    }
}

impl Widget for &mut App {
    fn render(self, area: Rect, buf: &mut Buffer) {
        self.layout = create_layout(area);

        if self.state.show_help {
            self.help_menu.render(area, buf, &mut self.state);
            return; // Don't render the rest of the UI when help is shown
        }

        // Regular UI rendering
        self.signal_list
            .render(self.layout.signal_list, buf, &mut self.state);
        self.waveform
            .render(self.layout.waveform, buf, &mut self.state);
        self.title_bar
            .render(self.layout.title, buf, &mut self.state);
        self.command_input
            .render(self.layout.command_bar, buf, &mut self.state);

        // Command result message if there is one
        if let Some((msg, is_error)) = &self.command_result {
            let status_style = if *is_error {
                Style::default().fg(Color::Red)
            } else {
                Style::default().fg(Color::Green)
            };

            // Create a temporary area just above the command bar for the result message
            let msg_area = Rect::new(
                self.layout.command_bar.x,
                self.layout.command_bar.y.saturating_sub(1),
                self.layout.command_bar.width.min(msg.len() as u16),
                1,
            );

            Paragraph::new(msg.clone())
                .style(status_style)
                .render(msg_area, buf);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::App;
    use insta::assert_snapshot;
    use ratatui::{backend::TestBackend, Terminal};

    #[test]
    fn test_render_app() {
        let mut app = App::default();
        let mut terminal = Terminal::new(TestBackend::new(80, 20)).unwrap();
        terminal
            .draw(|frame| frame.render_widget(&mut app, frame.area()))
            .unwrap();
        assert_snapshot!(terminal.backend());
    }
}
