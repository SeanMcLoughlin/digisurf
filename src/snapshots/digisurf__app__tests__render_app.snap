---
source: src/app.rs
expression: terminal.backend()
---
"┌──────────────────────────────────────────────────────────────────────────────┐"
"│DigiSurf | Time: 0 to 50 of 100                                               │"
"└──────────────────────────────────────────────────────────────────────────────┘"
"┌Signals───────┐⢲ ⢰⢲⢰⢲ ⢰⢲⢰⠒⢲⢰⢲⢰⠒⢲⢰⢲⢰⠒⢲⢰⢲ ⢰⢲⢰⢲ ⢰⢲⢰⠒⢲⢰⢲⢰⠒⢲⢰⢲ ⢰⢲⢰⢲ ⢰⢲⢰⢲ ⢰⢲⢰⠒⢲⢰⢲⢰⠒⢲ "
"│clk           │⠘⠒⠚⠘⠚⠘⠒⠚⠘⠚ ⠘⠚⠘⠚ ⠘⠚⠘⠚ ⠘⠚⠘⠒⠚⠘⠚⠘⠒⠚⠘⠚ ⠘⠚⠘⠚ ⠘⠚⠘⠒⠚⠘⠚⠘⠒⠚⠘⠚⠘⠒⠚⠘⠚ ⠘⠚⠘⠚ ⠘⠒"
"│              │⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⢲                                                   "
"│reset         │            ⠘⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒"
"│              │⢲           ⢰⢲           ⢰⢲          ⢰⠒⢲          ⢰⢲            "
"│data_valid    │⠘⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠚⠘⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠚⠘⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠚ ⠘⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠚⠘⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒⠒"
"│              │⣶⣀⣰⣰⣰⣰⣀⣰⣰⣰⣀⣰⣰⣰⣰⣀⣰⣰⣰⣰⣀⣰⣰⣰⣀⣰⣰⣰⣰⣀⣰⣰⣰⣀⣰⣰⣰⣰⣀⣰⣰⣰⣀⣰⣰⣰⣰⣀⣰⣰⣰⣰⣀⣰⣰⣰⣀⣰⣰⣰⣰⣀⣰⣀"
"│data          │⠛ ⠘⠘⠘⠘ ⠘⠘⠘ ⠘⠘⠘⠘ ⠘⠘⠘⠘ ⠘⠘⠘ ⠘⠘⠘⠘ ⠘⠘⠘ ⠘⠘⠘⠘ ⠘⠘⠘ ⠘⠘⠘⠘ ⠘⠘⠘⠘ ⠘⠘⠘ ⠘⠘⠘⠘ ⠘ "
"│              │⣰⠒⠒⢲⣰⠒⠒⢲⣰⠒⠒⢲⣰⠒⢲⣀⣰⠒⢲⣰⠒⠒⢲⣰⠒⠒⢲⣰⠒⠒⢲⣰⠒⠒⢲⣰⠒⢲⣀⣰⠒⢲⣀⣰⠒⢲⣰⠒⠒⢲⣰⠒⠒⢲⣰⠒⠒⢲⣰⠒⢲⣀⣰⠒"
"│tristate      │                                                                "
"│              │⣀    ⢀⣀⣀    ⢀⣀    ⢀⣀     ⢀⣀    ⢀⣀    ⢀⣀⣀    ⢀⣀    ⢀⣀     ⢀⣀     "
"│undefined     │⠘⠒⠒⠒⠒⠚ ⠘⠒⠒⠒⠒⠚⠘⠒⠒⠒⠒⠚⠘⠒⠒⠒⠒⠒⠚⠘⠒⠒⠒⠒⠚⠘⠒⠒⠒⠒⠚ ⠘⠒⠒⠒⠒⠚⠘⠒⠒⠒⠒⠚⠘⠒⠒⠒⠒⠒⠚⠘⠒⠒⠒⠒⠒"
"│              │                                                                "
"└──────────────┘                                                                "
"Normal──────────────────────────────────────────────────────────────────────────"
" ':' for command mode. :q, then <Enter> to quit. :h then <Enter> for help.      "
"                                                                                "
